### PR TITLE
fix(engine-core): correctly remap light DOM nested slot assignments

### LINKED ISSUE #3843: Nested light DOM slots don't render correctly
### Description

Light DOM slots are not rendering correctly when there are nested slots, [repro](https://stackblitz.com/edit/salesforce-lwc-1y8bjr?file=package.json):

```html
<template>
    <x-light-child>
        <div slot="namedSlot">Named Slot Content</div>
        <div>Unnamed Slot Content</div>
    </x-light-child>
</template>

// parent
```

```html
<template lwc:render-mode="light">
  <h2>Slotted content in Light DOM (note missing named slot)</h2>
  <x-leaf>
    <slot name="namedSlot"></slot>
  </x-leaf>
  <x-leaf>
    <slot></slot>
  </x-leaf>
</template>

// x-light-child
```

```html
<template>
  <slot></slot>
</template>

// x-leaf
```

In the above example, the slotted content "Named Slot Content" should render in the default slot for `x-leaf` but does not.

This is happening because for light DOM slots, the [children are directly returned](https://github.com/salesforce/lwc/blob/master/packages/%40lwc/engine-core/src/framework/api.ts#L261-L266) rather than [creating a `slot` element](https://github.com/salesforce/lwc/blob/master/packages/%40lwc/engine-core/src/framework/api.ts#L268-L272) like in synthetic shadow.

#### Root Cause

Each component that contains a `slot` will look up the slot's name in a slot map called `slotset` that determines what's slotted into the component.

This mapping is determined at runtime by traversing through the vnodes and their children.

When a new component containing a `slot` element is generated, it will look at the slot mapping from its parent to determine both the slot name and the content of what gets slotted.

Since light DOM slots directly return the content rather than the slot element, when there are nested slots, the mapping past the first level is missing and causes the component to render incorrectly.

### Details 

In the above example, the two `div` elements are [mapped to the slot mapping ](https://github.com/salesforce/lwc/blob/master/packages/%40lwc/engine-core/src/framework/rendering.ts#L834-L857):

```html
<template>
    <x-light-child>
        <div slot="namedSlot">Named Slot Content</div>
        <div>Unnamed Slot Content</div>
    </x-light-child>
</template>

```

The above will generate a slot map that looks something like this:

```javascript
{
    "nameSlot":  <div slot="namedSlot">Named Slot Content</div>, // This is for the named slot
    "": <div>Unnamed Slot Content</div> // This is for the default slot
}

// slotset
```

This map is used to [determine what will go in the slotted content inside](https://github.com/salesforce/lwc/blob/master/packages/%40lwc/engine-core/src/framework/api.ts#L210C33-L210C40) `x-light-child` (value of `slotset`):

```html
<template lwc:render-mode="light">
  <h2>Slotted content in Light DOM (note missing named slot)</h2>
  <x-leaf>
    <slot name="namedSlot"></slot>
  </x-leaf>
...
</template>

// x-light-child
```

Based on the `slotset` mapping, the children of `x-leaf` will be replaced with the vnode equivalent of `<div slot="namedSlot">Named Slot Content</div>`

Once the children have been allocated, they will be used to [determine the slot mapping of `x-leaf`](https://github.com/salesforce/lwc/blob/master/packages/%40lwc/engine-core/src/framework/rendering.ts#L827):

```javascript
{
    "nameSlot":  <div slot="namedSlot">Named Slot Content</div>,
}
```

When we finally get to `x-leaf`, since there is no default slot ("" entry in the `slotset`), the content of `x-leaf` is not rendered:

```html
<template>
  <slot></slot>
</template>

// x-leaf
```

The `slot` element needs an empty string entry to retrieve the children.

To demonstrate this, if we add `name="namedSlot"` the component will render correctly.

```html
<template>
  <slot name="namedSlot"></slot>
</template>
```

The extra `slot` element acts as a buffer to help set the default slot mapping.

I pushed up a [branch](https://github.com/salesforce/lwc/compare/jtu/lightdom-slots-broken?expand=1) to demonstrate a potential fix for the existing issue, although this change causes the karma tests to fail.

The branch also has the example described here to test with.

### PR DESCRIPTION
**Pull request by jmsjtu**
## Details
Fixes #3843

This PR fixes the issue of light DOM slots not mapping correctly when they are forwarded to another component.

Essentially, when a light DOM slot is passed as a child to another element it will return the children rather than create a slot element.

This can cause the slot mapping to be incorrect when the children have a `slot` attribute already.

For example:

```html
<x-light-child>
    <div slot="namedSlot">Named slot content</div>
    <div>Unnamed slot content</div>
</x-light-child>
```

```html
<template lwc:render-mode="light">
    <x-leaf>
        <slot></slot>
    </x-leaf>
</template>

// x-light-child
```

```html
<template>
    <slot></slot>
</template>

// x-leaf
```

In the above example, when the `slot` element is passed `x-leaf`, the children are passed directly (no `slot` element is created).

Since the `div` element has a `slot="namedSlot"` attribute, the slot mapping will attempt to map to a slot named "namedSlot" which doesn't exist and does not render.

This is especially an issue with native shadow DOM as in order to render the slot correctly, the `slot` attribute must be included on the element.

Note, this is only an issue when the root of the slotted content is a `slot`. In the above example, if the `slot` was wrapped in a `div` the content would render correctly.

ex:

```html
```html
<template lwc:render-mode="light">
    <x-leaf>
        <div>
            <slot></slot>
        </div>
    </x-leaf>
</template>
// x-light-child
```

This PR resolves the issue by replacing the `slot` attribute on the children with the correct `slot` attribute mapping.

In above example, the div `slot` attribute will be removed on the `div` so that it can render properly.

## Does this pull request introduce a breaking change?

* ✅ No, it does not introduce a breaking change.

## Does this pull request introduce an observable change?

* ⚠️ Yes, it does include an observable change.

Any light DOM elements that contain nested slots will have the `slot` attribute changed, this may impact some users but I don't anticipate it having a large impact.

## GUS work item
W-14356488

### CODE REVIEW

File: packages/@lwc/engine-core/src/framework/api.ts
Comment (jmsjtu) [inline packages/@lwc/engine-core/src/framework/api.ts:269]: Creating a copy instead of overwriting the attributes here since everything in `VElementData` should be read only.

File: packages/@lwc/integration-karma/test/light-dom/scoped-slot/index.spec.js
Comment (jmsjtu) [inline packages/@lwc/integration-karma/test/light-dom/scoped-slot/index.spec.js:64]: Since the light DOM slot mapping changes based on the `slot` element, the `slot` attribute was removed from this element. 

It still renders in the same position though.

File: packages/@lwc/integration-karma/test/light-dom/slotting/index.spec.js
Comment (jmsjtu) [inline packages/@lwc/integration-karma/test/light-dom/slotting/index.spec.js:153]: These tests just ensure that when a light DOM slot is passed to a shadow DOM element the slot mapping is preserved.

It swaps the first and last elements originally being slotted.

File: packages/@lwc/engine-core/src/framework/api.ts
Comment (divmain) [reply packages/@lwc/engine-core/src/framework/api.ts:269]: Might be worth adding a reminder to the comment here that `VNodeData` might be shared between `VNodes` as an optimization, which is why `vnode.data` is unsafe to mutate.

**Review by divmain — COMMENTED** [review]
This seems reasonable to me, though I'd wait for @abdulsattar or @nolanlawson to weigh in before merging.

File: packages/@lwc/engine-core/src/framework/api.ts
Comment (nolanlawson) [inline packages/@lwc/engine-core/src/framework/api.ts:255]: This can be done after this PR (to avoid confusion), but we should probably refactor the entire `s` function to reduce the cyclomatic complexity. It's really hard to follow all these nested if/elses.

File: packages/@lwc/integration-karma/test/light-dom/slotting/index.spec.js
Comment (nolanlawson) [inline packages/@lwc/integration-karma/test/light-dom/slotting/index.spec.js:137]: I recommend avoiding `innerHTML` entirely since it's buggy between native/synthetic shadow. Ideally we should have the same tests here for both native and synthetic.

**Review by abdulsattar — APPROVED** [review]
LGTM once you address the other concerns (and the one with scoped slots).

**Comment by jmsjtu** [comment]
Closing in favor of #3883

**Pull request closed without merging**
